File: src/util/rtp.ts
import { bot } from ':src/bot.ts';
import { HOME_GUILD_ID, RTP_ROLE_ID } from ':src/env.ts';
import { log } from ':util/logger.ts';
import { Member } from 'discordeno';

/**
 * Checks for all home guild's members with the configured ready-to-play role.
 * @returns an array of all members with the role
 */
export const getRTPMembers = async (): Promise<Member[]> => {
    const allMembers = await bot.helpers.getMembers(HOME_GUILD_ID, {});
    const rtpMembers = allMembers
        .array()
        .filter((member) => member.roles.includes(RTP_ROLE_ID));

    return rtpMembers;
};

/**
 * Adds the configured ready-to-play role to a given member.
 * @param member to add the role to
 */
export const addMemberToRTP = async (member: Member) => {
    await bot.helpers.addRole(HOME_GUILD_ID, member.id, RTP_ROLE_ID);
    log('Added member to RTP');
};

/**
 * Removes the configured ready-to-play role from a given member.
 * @param member to remove the role from
 */
export const removeMemberFromRTP = async (member: Member) => {
    await bot.helpers.removeRole(HOME_GUILD_ID, member.id, RTP_ROLE_ID);
    log('Removed member from RTP');
};

File: src/util/rtp.rs
use poise::serenity_prelude::{Context, Member};

use crate::{env::Environment, types::error::Error};

/// Count how many members have the ready role.
pub async fn count(context: &Context) -> Result<usize, Error> {
    let env = Environment::instance();

    let members = context
        .http
        .get_guild_members(env.home_guild_id, None, None)
        .await?;

    let ready_member_count = members
        .iter()
        .filter(|member| {
            member
                .roles
                .iter()
                .any(|role_id| role_id == &env.ready_role_id)
        })
        .count();

    Ok(ready_member_count)
}

/// Add the ready role to a member.
pub async fn add(member: &mut Member, context: &Context) -> Result<(), Error> {
    member
        .add_role(context, Environment::instance().ready_role_id)
        .await?;

    log::debug!("Added {} to ready role", member.user.tag());

    Ok(())
}

/// Remove the ready role from a member.
pub async fn remove(member: &mut Member, context: &Context) -> Result<(), Error> {
    member
        .remove_role(context, Environment::instance().ready_role_id)
        .await?;

    log::debug!("Removed {} from ready role", member.user.tag());

    Ok(())
}
